### PR TITLE
feat(core): Flexible connection resolution strategy when using read-replicas

### PR DESCRIPTION
**Pull request by asiFarran**
Allows developers working with read-replicas the flexibility to:

- Set the default resolution strategy for read operations (resolve to a read-replica or write connection)
- Be explicit on a per-query basis by extending `FindOptions` and `CountOptions` with a `connectionType` property

### CODE REVIEW

**Comment by B4nan** [comment]
please remove all unrelated code style changes, and the yarn log file

**Comment by B4nan** [comment]
Note that we need to keep supporting the string constants (otherwise this would be a breaking change on type level), so remove all of those changes too.

**Comment by asiFarran** [comment]
> Note that we need to keep supporting the string constants (otherwise this would be a breaking change on type level), so remove all of those changes too.

I restored the string constants but did wrap them to a union-type so to make signatures shorter & 'safer'

File: packages/knex/src/AbstractSqlDriver.ts [hunk 1]
```diff
@@ -620,8 +618,9 @@ export abstract class AbstractSqlDriver<C extends AbstractSqlConnection = Abstra
   }
 
   /** @internal */
-  createQueryBuilder<T extends AnyEntity<T>>(entityName: string, ctx?: Transaction<Knex.Transaction>, write?: boolean, convertCustomTypes?: boolean): QueryBuilder<T> {
-    const qb = new QueryBuilder<T>(entityName, this.metadata, this, ctx, undefined, write ? 'write' : 'read');
+  createQueryBuilder<T extends AnyEntity<T>>(entityName: string, ctx?: Transaction<Knex.Transaction>, prefferedConnectionType?: ConnectionType, convertCustomTypes?: boolean): QueryBuilder<T> {
+    const connectionType = this.resolveConnectionType({ ctx, connectionType: prefferedConnectionType });
```
Comment (B4nan) [inline packages/knex/src/AbstractSqlDriver.ts:622]: typo in `preferred`
```suggestion
    const connectionType = this.resolveConnectionType({ ctx, connectionType: preferredConnectionType });
```

File: tests/EntityManager.mysql.test.ts [hunk 2]
```diff
@@ -1940,50 +1938,212 @@ describe('EntityManagerMySql', () => {
     process.env.FORCE_COLOR = '0';
   });
 
-  test('read replicas', async () => {
-    const mock = mockLogger(orm, ['query']);
+  describe('read-replicas',  () => {
```
Comment (B4nan) [inline tests/EntityManager.mysql.test.ts:1941]: let's move this to `tests/features/read-replicas`

**Review by B4nan — CHANGES_REQUESTED** [review]
looking good, just two nits

File: tests/features/read-replicas.test.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  afterEach(() => {
+    orm.config.set('debug', false);
+    Author2Subscriber.log.length = 0;
+    EverythingSubscriber.log.length = 0;
+    FlushSubscriber.log.length = 0;
+    Test2Subscriber.log.length = 0;
+  });
+
+  describe('when preferReadReplicas is true (default behaviour)', () => {
+      test('will prefer replicas for read operations outside a transaction', async () => {
```
Comment (B4nan) [inline tests/features/read-replicas.test.ts:25]: one last nit: bad indentation all over the new file

**Pull request merged**